plugins/CoreConsole/Commands/DevelopmentSyncUITestScreenshots.php

```diff
@@ -73,42 +71,166 @@ protected function execute(InputInterface $input, OutputInterface $output)
+
+        $downloadTo = PIWIK_DOCUMENT_ROOT . "/";
+        if (empty($plugin)) {
+            $downloadTo .= "tests/UI/expected-ui-screenshots/";
+        } else {
+            $downloadTo .= "plugins/$plugin/tests/UI/expected-ui-screenshots/";
+        }
+        if(is_dir($downloadTo)) {
+            return $downloadTo;
+        }
+        $downloadTo = str_replace("tests/", "Test/", $downloadTo);
+        if(is_dir($downloadTo)) {
+            return $downloadTo;
+        }
+
+        throw new \Exception("Download to path could not be found: $downloadTo");
+    }
+
+    protected function getDiffviewerContent(InputInterface $input)
+    {
+        $this->urlBase = $this->getUrlBase($input);
+        $diffviewerUrl = $this->getDiffviewerUrl($this->urlBase);
+
+        $diffviewer = $this->downloadDiffviewer($diffviewerUrl);
+        if($diffviewer) {
+            return $diffviewer;
+        }
+
+        // Maybe this is a Premium Piwik PRO plugin...
+        return $this->getDiffviewContentForPrivatePlugin($input);
+    }
+
+    protected function getDiffviewContentForPrivatePlugin(InputInterface $input)
+    {
+        $httpUser = $input->getOption('http-user');
+        $httpPassword = $input->getOption('http-password');
+        if (empty($httpUser) || empty($httpPassword)) {
+            Log::info("--http-user and --http-password was not specified, skip download of private plugins screenshots.");
+            return;
+        }
+
+        // Attempt to download from protected/ artifacts...
+        $this->urlBase = str_replace("builds-artifacts.piwik.org/", "builds-artifacts.piwik.org/protected/", $this->urlBase);
+        $diffviewerUrl = $this->getDiffviewerUrl($this->urlBase);
+
+        return $this->downloadDiffviewer($diffviewerUrl, $httpUser, $httpPassword);
+    }
+
+    /**
+     * @return string
+     */
+    protected function getDiffviewerUrl($urlBase)
+    {
+        return $urlBase . "/screenshot-diffs/diffviewer.html";
+    }
+
+    protected function downloadDiffviewer($urlDiffviewer, $httpUsername = false, $httpPassword = false)
+    {
+        $responseExtended = Http::sendHttpRequest(
+            $urlDiffviewer,
+            $timeout = 60,
+            $userAgent = null,
+            $destinationPath = null,
+            $followDepth = 0,
+            $acceptLanguage = false,
+            $byteRange = false,
+            $getExtendedInfo = true,
+            $httpMethod = 'GET',
+            $httpUsername,
+            $httpPassword
+        );
+        $httpStatus = $responseExtended['status'];
+        if ($httpStatus == '200') {
+            $diffviewer = str_replace('&', '&amp;', $responseExtended['data']);
+            return $diffviewer;
+        }
+
+        Log::info("Could not download the diffviewer from $urlDiffviewer - Got HTTP status " . $httpStatus);
+        if($httpStatus == '401') {
+            Log::warning("HTTP AUTH username and password are not valid.");
+        }
+        return false;
```

Comment:
If `return false` is used to signal an error then exceptions should be used instead.